wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointMonitorImpl.java

```diff
@@ -0,0 +1,238 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.plugin.customendpoint;
+
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiFunction;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.rds.RdsClient;
+import software.amazon.awssdk.services.rds.model.DBClusterEndpoint;
+import software.amazon.awssdk.services.rds.model.DescribeDbClusterEndpointsResponse;
+import software.amazon.awssdk.services.rds.model.Filter;
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.PluginService;
+import software.amazon.jdbc.util.CacheMap;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.RdsUtils;
+import software.amazon.jdbc.util.StringUtils;
+import software.amazon.jdbc.util.telemetry.TelemetryCounter;
+import software.amazon.jdbc.util.telemetry.TelemetryFactory;
+
+/**
+ * The default custom endpoint monitor implementation. This class uses a background thread to monitor a given custom
+ * endpoint for custom endpoint information and future changes to the custom endpoint.
+ */
+public class CustomEndpointMonitorImpl implements CustomEndpointMonitor {
+  private static final Logger LOGGER = Logger.getLogger(CustomEndpointPlugin.class.getName());
+  private static final String TELEMETRY_ENDPOINT_INFO_CHANGED = "customEndpoint.infoChanged.counter";
+
+  // Keys are custom endpoint URLs, values are information objects for the associated custom endpoint.
+  protected static final CacheMap<String, CustomEndpointInfo> customEndpointInfoCache = new CacheMap<>();
+  protected static final long CUSTOM_ENDPOINT_INFO_EXPIRATION_NANO = TimeUnit.MINUTES.toNanos(5);
+
+  protected final AtomicBoolean stop = new AtomicBoolean(false);
+  protected final RdsUtils rdsUtils = new RdsUtils();
+  protected final RdsClient rdsClient;
+  protected final HostSpec customEndpointHostSpec;
+  protected final String endpointIdentifier;
+  protected final Region region;
+  protected final long refreshRateNano;
+
+  protected final PluginService pluginService;
+  // TODO: is it problematic for each monitor thread to have its own executor service?
```

Comment:
It's not a problem.

Reply:
Okay, I had read somewhere that creating many executor services consumes more system resources and might lead to performance degradation, and I've noticed we've done this elsewhere, but it hasn't seemed to cause performance trouble so far.